#160451503 Add functionality to rate articles

kampkelly opened the pull request:
#### What does this PR do?
This pull request adds the functionality to rate an article.

#### Description of Task to be completed?
- Write tests for rating articles
- Write the code to rate an article

#### How should this be manually tested?
After cloning the project, run `npm install`, then `npm run watch` to start the server. Using postman and as a signed in user, on sending an integer or float as 'rating' to the rate API endpoint along with the correct url of an article, then an article should be rated.

#### Any background context you want to provide?
None
#### What are the relevant pivotal tracker stories?
#160451503
#### Questions:```
None

kampkelly committed:
feature(rate-articles): write rating tests

This commit cretates the model and migration for the ratings. The test cases hav also been written in this commit.

Files changed:
- server/tests/rate.test.js
- server/tests/rateModels.test.js
- server/controllers/rateController.js
- server/models/rate.js
- server/migrations/20181008153946-create-rate.js
- server/routes/api/article.js

Start #160451503

kampkelly committed:
feature(rate-articles): write rating fucntionality

This commit rates an article. When a valid number or float is sent to the rate api endpoint, an article will be rated and its average computed. This average consists of all the ratings for an article and it is stored int the articles table.
A test to check if the rating sent is a valid number or float has also been added.

Files changed:
- server/tests/article.test.js
- server/tests/rate.test.js
- server/controller./rateController.js
- server/models/article.js
- server/migrations/20181003095441-create-article.js

Finishes #160451503

kampkelly commented on server/tests/article.test.js:
```diff
@@ -15,9 +15,9 @@ const user = {
+  title: faker.lorem.sentence(),
+  body: faker.lorem.paragraph(),
+  description: faker.lorem.sentences()
 };
```
I changed the random words generated by faker to be generated from lorem sentences. This is because the tests were failing since the random words may occasionally be less than the required minimum length.

kampkelly commented on server/controllers/rateController.js:
```diff
@@ -0,0 +1,79 @@
+              return Rate.create({
+                userId: req.user.id,
+                articleId: article.dataValues.id,
+                rating: req.body.rating
```
I put this in a single line but the team's eslint settings with "--fix" auto corrects it, placing them in separate lines.

kampkelly committed:
eature(rate-articles): set range for rating

This commit adds a range for rating an article. Test has also been added for the rating range.

Finishes #160451503

oladimillion commented:
LGTM

udochukwu reviewed (commented):
LGTM

slimsolz reviewed (commented):
LGTM

onwuvic reviewed (commented):
LGTM

samdiano commented:
LGTM

samdiano reviewed (commented):
LGTM

s-akinrele commented on server/controllers/rateController.js:
```diff
@@ -0,0 +1,85 @@
+  * @returns {object} - status, message and articles details
+  */
+  static rateArticle(req, res) {
+    if (Number.isNaN(Number(req.body.rating)) !== true) {
```
extract this into another method. or use Joi.. Also refactor your implementation

s-akinrele commented on server/controllers/rateController.js:
```diff
@@ -0,0 +1,85 @@
+        Article.findOne({
+          where: { slug: req.params.slug }
+        }).then((article) => {
+          if (article !== null) {
```
you dont need `!==null` .. `if(article)`

s-akinrele commented on server/controllers/rateController.js:
```diff
@@ -0,0 +1,85 @@
+                    articleId: article.dataValues.id,
+                    rating: req.body.rating
+                  });
+                }
```
you if else here instead...

s-akinrele commented on server/tests/rateModels.test.js:
```diff
@@ -0,0 +1,78 @@
+// import faker from 'faker';
```
remove commented code.

s-akinrele commented on server/tests/rateModels.test.js:
```diff
@@ -0,0 +1,78 @@
+    expect(rate.articleId).to.equal(validRate.articleId);
+  });
+
+  describe('Rating validation', () => {
```
can this be `validate ratings` instead.

s-akinrele reviewed (changes requested):
I had some comments

kampkelly committed:
feature(rate-articles): refactor code

This commit refactors the code for easy readability.
The validations for rating an article has been put in a separate middleware.
The computing of the averages of the rating has been put in an helper file.

Finishes #160451503

kampkelly committed:
feature(rate-articles): set range for rating

This commit fixes eslint —fix issue.

Finishes #160451503

kampkelly committed:
feature(rate-articles): refactor code

This commit refactors part of the code.

Finishes #160451503

kampkelly committed:
feature(rate-articles): refactor more code

This commit refactors and removes unused code.

Finishes #160451503
